mdiff: panic (slice bounds out of range) when using Unify()

I was using mdiff to compute a diff of a big file, and hit a panic. It seems to only happen when using Unify(), or possibly Unify() just alters the state to expose the crash, not sure.

Reproduction recipe:

```
git clone https://gist.github.com/107d080ac77504da650db60afa234c1a.git repro
cd repro
go run .
```

Result:

```
panic: runtime error: slice bounds out of range [:-1]

goroutine 1 [running]:
github.com/creachadair/mds/mdiff.UnifyChunks({0xc000120488, 0x69, 0x8f})
	/home/dave/go/pkg/mod/github.com/creachadair/mds@v0.15.2/mdiff/mdiff.go:254 +0x67c
github.com/creachadair/mds/mdiff.(*Diff).Unify(...)
	/home/dave/go/pkg/mod/github.com/creachadair/mds@v0.15.2/mdiff/mdiff.go:176
main.main()
	/home/dave/hack/psl/tools/fuck/107d080ac77504da650db60afa234c1a/main.go:23 +0x16a
exit status 2
```

Unfortunately the diff inputs are large, so definitely not a minimal repro. But hopefully is enough to pinpoint the source. I can also go digging, but I don't know mdiff at all and I'm in a hurry right now so thought I'd at least write it down for later :)

--- a/mdiff/mdiff.go
+++ b/mdiff/mdiff.go
@@ -247,7 +247,7 @@ func UnifyChunks(cs []*Chunk) []*Chunk {
 		// original script edits.
 		if lap > 0 {
 			if end.Op == slice.OpEmit { // last has post-context
-				if lap == len(end.X) { // remove the whole edit
+				if lap >= len(end.X) { // remove the whole edit
 					last.Edits = last.Edits[:len(last.Edits)-1]
 					end = slice.PtrAt(last.Edits, -1)
 				} else {
@@ -257,7 +257,7 @@ func UnifyChunks(cs []*Chunk) []*Chunk {
 				last.LEnd -= lap
 				last.REnd -= lap
 			} else if start.Op == slice.OpEmit { // start has pre-context
-				if lap == len(start.X) { // remove the whole edit
+				if lap >= len(start.X) { // remove the whole edit
 					c.Edits = c.Edits[1:]
 					start = slice.PtrAt(c.Edits, 0)
 				} else {

--- a/mdiff/mdiff_test.go
+++ b/mdiff/mdiff_test.go
@@ -84,6 +84,24 @@ func TestDiff(t *testing.T) {
 	})
 }
 
+func TestRegression(t *testing.T) {
+	t.Run("#12", func(t *testing.T) {
+		const contextWindow = 3
+
+		// Produce a chunk with a longer overlap than the size of the context
+		// window.  Without the fix, this will trigger a panic in unification.
+		lhs := mstr.Lines("X\nY\nZ\n\na\nb\nc")
+		rhs := mstr.Lines("X\nY\nZ\n\n\na\n\nb\n\nc\n\n")
+
+		d := mdiff.New(lhs, rhs)
+		t.Log("-- Before context")
+		logChunks(t, d.Chunks)
+		d.AddContext(contextWindow).Unify()
+		t.Log("-- After unification")
+		logChunks(t, d.Chunks)
+	})
+}
+
 func TestNoAlias(t *testing.T) {
 	// The documentation promises that adding context and unifying does not
 	// disturb the original edit sequence.

--- a/mdiff/mdiff_test.go
+++ b/mdiff/mdiff_test.go
@@ -90,8 +90,8 @@ func TestRegression(t *testing.T) {
 
 		// Produce a chunk with a longer overlap than the size of the context
 		// window.  Without the fix, this will trigger a panic in unification.
-		lhs := mstr.Lines("X\nY\nZ\n\na\nb\nc")
-		rhs := mstr.Lines("X\nY\nZ\n\n\na\n\nb\n\nc\n\n")
+		lhs := lines("X", "Y", "Z", "", "a", "b", "c")
+		rhs := lines("X", "Y", "Z", "", "", "a", "", "b", "", "c", "")
 
 		d := mdiff.New(lhs, rhs)
 		t.Log("-- Before context")
@@ -300,3 +300,5 @@ func logChunks(t *testing.T, cs []*mdiff.Chunk) {
 		}
 	}
 }
+
+func lines(ss ...string) []string { return ss }

